File: mobile-app/lib/services/match_service.dart
import 'package:ludo_legends/services/api_client.dart';
import 'package:ludo_legends/core/config/api_config.dart';
import 'package:ludo_legends/models/match.dart';

class MatchService {
  final ApiClient _api;

  MatchService({ApiClient? api}) : _api = api ?? ApiClient();

  Future<Match> submitMatch(String tournamentId, String screenshotUrl, {String? resultNotes}) async {
    final response = await _api.dio.post('${ApiConfig.matches}/submit', data: {
      'tournament_id': tournamentId,
      'screenshot_url': screenshotUrl,
      if (resultNotes != null) 'result_notes': resultNotes,
    });
    return Match.fromJson(response.data);
  }

  Future<List<Match>> getMyMatches({int page = 1, int perPage = 20}) async {
    final response = await _api.dio.get('${ApiConfig.matches}/my', queryParameters: {
      'page': page,
      'per_page': perPage,
    });
    return (response.data as List).map((m) => Match.fromJson(m)).toList();
  }

  Future<List<Match>> getPendingMatches({int page = 1, int perPage = 20}) async {
    final response = await _api.dio.get('${ApiConfig.matches}/pending', queryParameters: {
      'page': page,
      'per_page': perPage,
    });
    return (response.data as List).map((m) => Match.fromJson(m)).toList();
  }

  Future<Match> verifyMatch(String matchId, String action, {String? winnerId, String? score, double? prizeAwarded, String? rejectionReason}) async {
    final response = await _api.dio.post('${ApiConfig.matches}/$matchId/verify', data: {
      'action': action,
      if (winnerId != null) 'winner_id': winnerId,
      if (score != null) 'score': score,
      if (prizeAwarded != null) 'prize_awarded': prizeAwarded,
      if (rejectionReason != null) 'rejection_reason': rejectionReason,
    });
    return Match.fromJson(response.data);
  }
}

File: backend-api/app/services/match_service.py
"""Match service — submit, verify, list matches."""

import uuid
from datetime import datetime, timezone
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select, func
from app.models.match import Match, MatchResult, MatchStatus
from app.models.user import User
from app.services.wallet_service import WalletService
from app.models.wallet import TransactionType


class MatchService:
    def __init__(self, db: AsyncSession):
        self.db = db

    async def submit_match(
        self,
        user_id: str,
        tournament_id: str,
        screenshot_url: str,
        result_notes: str | None = None,
    ) -> Match:
        """Submit a match result for admin verification."""
        match = Match(
            id=str(uuid.uuid4()),
            tournament_id=tournament_id,
            user_id=user_id,
            screenshot_url=screenshot_url,
            result_notes=result_notes,
            status=MatchStatus.SUBMITTED,
            submitted_at=datetime.now(timezone.utc),
        )
        self.db.add(match)
        await self.db.flush()
        await self.db.refresh(match)
        return match

    async def verify_match(
        self,
        match_id: str,
        action: str,
        winner_id: str | None = None,
        score: str | None = None,
        prize_awarded: float = 0,
        rejection_reason: str | None = None,
    ) -> Match:
        """Admin verify a match — approve with prize or reject."""
        result = await self.db.execute(select(Match).where(Match.id == match_id))
        match = result.scalar_one_or_none()
        if not match:
            raise ValueError("Match not found")
        if match.status != MatchStatus.SUBMITTED:
            raise ValueError("Match already verified")

        if action == "approve":
            match.status = MatchStatus.VERIFIED
            match.verified_at = datetime.now(timezone.utc)

            match_result = MatchResult(
                id=str(uuid.uuid4()),
                match_id=match_id,
                winner_id=winner_id,
                score=score,
                prize_awarded=prize_awarded,
            )
            self.db.add(match_result)

            if winner_id and prize_awarded > 0:
                wallet_svc = WalletService(self.db)
                await wallet_svc.credit(
                    winner_id, prize_awarded, TransactionType.PRIZE,
                    match_id, f"Prize for match {match_id[:8]}"
                )

            if winner_id:
                winner_result = await self.db.execute(select(User).where(User.id == winner_id))
                winner = winner_result.scalar_one_or_none()
                if winner:
                    winner.total_wins += 1
                    winner.total_earnings += prize_awarded
        else:
            match.status = MatchStatus.REJECTED
            if rejection_reason:
                match.rejection_reason = rejection_reason

        await self.db.flush()
        await self.db.refresh(match)
        return match

    async def get_user_matches(
        self, user_id: str, page: int = 1, per_page: int = 20
    ) -> tuple[list[Match], int]:
        """Get paginated matches for a user."""
        offset = (page - 1) * per_page
        result = await self.db.execute(
            select(Match)
            .where(Match.user_id == user_id)
            .order_by(Match.created_at.desc())
            .offset(offset)
            .limit(per_page)
        )
        matches = list(result.scalars().all())

        count_result = await self.db.execute(
            select(func.count(Match.id)).where(Match.user_id == user_id)
        )
        total = count_result.scalar() or 0
        return matches, total

    async def get_pending_matches(
        self, page: int = 1, per_page: int = 20
    ) -> tuple[list[Match], int]:
        """Get paginated pending matches for admin review."""
        offset = (page - 1) * per_page
        result = await self.db.execute(
            select(Match)
            .where(Match.status == MatchStatus.SUBMITTED)
            .order_by(Match.submitted_at.asc())
            .offset(offset)
            .limit(per_page)
        )
        matches = list(result.scalars().all())

        count_result = await self.db.execute(
            select(func.count(Match.id)).where(Match.status == MatchStatus.SUBMITTED)
        )
        total = count_result.scalar() or 0
        return matches, total
